File: 2023/bun/src/day-04/card.ts
import { z } from 'zod';

type Card = {
  id: number;
  winningNumbers: number[];
  numbers: number[];
};

export const getWinningMatches = (card: Card) => {
  return card.numbers.filter((x) => card.winningNumbers.includes(x));
};

export const buildCardFromLine = (line: string) => {
  const regex = line.match(/Card\s+([0-9]+)/);

  const rawCards = line.split(':')[1].split('|');

  const winningNumbers = z
    .array(z.coerce.number())
    .parse(rawCards[0].split(' ').filter((x) => x !== ''));

  const numbers = z
    .array(z.coerce.number())
    .parse(rawCards[1].split(' ').filter((x) => x !== ''));

  return {
    id: z.coerce.number().parse(regex?.[1]),
    numbers,
    winningNumbers,
  } satisfies Card;
};

export function processScratchCards(cards: Card[]): Card[] {
  return Array.from(
    cards
      .reduce(
        (acc, card, index, cards) => {
          // Get the number of matches for the current card
          const matches = getWinningMatches(card).length;

          if (matches === 0) return acc;

          // Get the number of copies of the current card
          const currentNumberOfCopies = acc.get(card) ?? 0;

          // Get the cards that will be won by the current card
          const copies = cards.slice(index + 1, index + matches + 1);

          // Iterate over the cards that will be won by the current card
          copies.forEach((card) => {
            // Get the number of copies of the card that will be won by the current card
            const numberOfCopies = acc.get(card) ?? 0;

            // Add the number of copies of the card that will be won by the current card to the accumulator
            acc.set(card, numberOfCopies + currentNumberOfCopies);
          });

          return acc;
        },
        new Map(cards.map((card) => [card, 1]))
      )
      .entries()
  )
    .map(([card, count]) => {
      return Array.from({
        length: count,
      }).map(() => ({ ...card }));
    })
    .flat();
}

File: 2023/rust/day-04/src/card.rs
use std::collections::HashMap;

#[derive(PartialEq, Eq, Hash, Clone, Debug)]
pub struct Card {
    pub id: i32,
    winning_numbers: Vec<i32>,
    numbers: Vec<i32>,
}

impl Card {
    pub fn new(line: &str) -> Card {
        let card_sections = line.split(':').collect::<Vec<&str>>();
        let id = card_sections[0]
            .split(' ')
            .last()
            .unwrap()
            .parse::<i32>()
            .unwrap();

        let raw_cards = card_sections[1].split(" | ").collect::<Vec<&str>>();

        let winning_numbers = raw_cards[0]
            .split(' ')
            .filter(|x| !x.is_empty())
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        let numbers = raw_cards[1]
            .split(' ')
            .filter(|x| !x.is_empty())
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        Card {
            id,
            winning_numbers,
            numbers,
        }
    }

    pub fn get_winning_matches(&self) -> Vec<i32> {
        self.numbers
            .iter()
            .filter(|x| self.winning_numbers.contains(x))
            .map(|x| *x)
            .collect::<Vec<i32>>()
    }
}
pub struct ScratchCards {
    cards: Vec<Card>,
    map: HashMap<i32, Card>,
}

impl ScratchCards {
    pub fn new(cards: Vec<Card>) -> ScratchCards {
        let mut map = HashMap::new();

        cards.iter().for_each(|card| {
            map.insert(card.id, card.clone());
        });

        ScratchCards { cards, map }
    }

    pub fn get_copies(&self) -> Vec<&Card> {
        let mut initial_map = HashMap::new();

        self.cards.iter().for_each(|card| {
            initial_map.insert(card.id, 1);
        });

        let result = self
            .cards
            .iter()
            .enumerate()
            .fold(initial_map, |mut map, (index, card)| {
                let matches = card.get_winning_matches();
                let start_of_slice = index + 1;
                let end_of_slice = start_of_slice + matches.len();

                match matches.len() {
                    0 => (),
                    _ => {
                        let current_number_of_copies = *map.entry(card.id).or_insert(0);

                        self.cards[start_of_slice..end_of_slice]
                            .iter()
                            .for_each(|card| {
                                let number_of_copies = map.entry(card.id).or_insert(0);
                                *number_of_copies += current_number_of_copies;
                            })
                    }
                }

                map
            })
            .iter()
            .map(|(card_id, count)| {
                let card = self.map.get(card_id).unwrap();
                println!("card: {}, count: {}", card_id, count);
                // copy card count times
                let mut copies = Vec::new();

                for _ in 0..*count {
                    copies.push(card);
                }

                copies
            })
            .flatten()
            .collect::<Vec<&Card>>();

        result
    }
}
